### DISCUSSION TITLE
Google OAuth with existing user throws Validation Error

### DISCUSSION THREAD

**Discussion by n-bravo** [feathersjs/feathers#3231] [Q&A] [answered]
Hello Everyone,

Currently I'm adding Google OAuth to an application with Featherjs 4.5 following the official docs and everything works smoothly with a new user with no existing email on the database.

But for a existing user with null googleId, the OAuth throws Validation Error, refusing to patch the user by just adding the missing fields (like googleId and profilePicture).

Is there a way to allow this patch behavior with the Custom OAuthStrategy?

Thank you in advance!

**Comment by daffl** [comment]
The default oAuth strategy looks up the existing entity by `googleId`. How did you customise it to change that behaviour?

**Reply by n-bravo** [reply, marked as answer]
Hello @daffl ,

I managed to solve this issue by patching the googleId of the existing user in a beforeHook of the authentication service.

The hook looks likes this:
```
const authenticationHooks = {
  before: {
    create: [async (context) => {
      if (context.data.strategy === 'google' && context.data.id_token) {
        const data = (await context.app.service('users').find({query: {
          email: context.data.id_token.payload.email,
          googleId: null
        }, $limit: 1, $select: ['id']})).data;
        if (data.length > 0) {
          const userId = data[0].id;
          await context.app.service('users').patch(userId, {googleId: context.data.id_token.payload.sub});
        }
      }
      return context;
      
    }]
  }
}
```

I had to adapt the beforeHooks of create and patch of the `users` service to handle local calls (`context.params.provider === undefined`).

Hope this solution helps everyone that faces the same issue.
Any improvements to this implementation are welcome.

**Reply by daffl** [reply]
That definitely works. You could also [customise](https://feathersjs.com/api/authentication/oauth.html#customization) the [getEntityQuery](https://feathersjs.com/api/authentication/oauth.html#getentityquery-profile-params) similar to this:

```ts
class MyGoogleStrategy extends OAuthStrategy {
  getEntityQuery (profile, params) {
    return {
      $or: [{
        [`${this.name}Id`]: profile.sub || profile.id
      }, {
        email: profile.email
      }]
    }
  }
}
```
